staging/src/k8s.io/cloud-provider/controllers/service/controller_test.go

```diff
@@ -1366,71 +1372,23 @@ func TestNeedsUpdate(t *testing.T) {
 	}, {
 		testName: "If service IPFamilies from single stack to dual stack",
 		updateFn: func() {
-			protocol := "http"
-			oldSvc = &v1.Service{
-				ObjectMeta: metav1.ObjectMeta{
-					Name:      "tcp-service",
-					Namespace: "default",
-				},
-				Spec: v1.ServiceSpec{
-					Ports: []v1.ServicePort{{
-						Port:        80,
-						Protocol:    v1.ProtocolTCP,
-						TargetPort:  intstr.Parse("22"),
-						AppProtocol: &protocol,
-					}},
-					IPFamilies: []v1.IPFamily{v1.IPv4Protocol},
-					Type:       v1.ServiceTypeLoadBalancer,
-				},
-			}
+			oldSvc = newService("tcp-service", v1.ServiceTypeLoadBalancer, tweakAddPorts(v1.ProtocolTCP, 22), tweakAddAppProtocol("http"), tweakAddIPFamilies([]v1.IPFamily{v1.IPv4Protocol}))
 			newSvc = oldSvc.DeepCopy()
 			newSvc.Spec.IPFamilies = []v1.IPFamily{v1.IPv4Protocol, v1.IPv6Protocol}
 		},
 		expectedNeedsUpdate: true,
 	}, {
 		testName: "If service IPFamilies from dual stack to single stack",
 		updateFn: func() {
-			protocol := "http"
-			oldSvc = &v1.Service{
-				ObjectMeta: metav1.ObjectMeta{
-					Name:      "tcp-service",
-					Namespace: "default",
-				},
-				Spec: v1.ServiceSpec{
-					Ports: []v1.ServicePort{{
-						Port:        80,
-						Protocol:    v1.ProtocolTCP,
-						TargetPort:  intstr.Parse("22"),
-						AppProtocol: &protocol,
-					}},
-					IPFamilies: []v1.IPFamily{v1.IPv4Protocol, v1.IPv6Protocol},
-					Type:       v1.ServiceTypeLoadBalancer,
-				},
-			}
+			oldSvc = newService("tcp-service", v1.ServiceTypeLoadBalancer, tweakAddPorts(v1.ProtocolTCP, 22), tweakAddAppProtocol("http"), tweakAddIPFamilies([]v1.IPFamily{v1.IPv4Protocol, v1.IPv6Protocol}))
 			newSvc = oldSvc.DeepCopy()
 			newSvc.Spec.IPFamilies = []v1.IPFamily{v1.IPv4Protocol}
 		},
 		expectedNeedsUpdate: true,
 	}, {
 		testName: "If service IPFamilies not change",
 		updateFn: func() {
-			protocol := "http"
-			oldSvc = &v1.Service{
-				ObjectMeta: metav1.ObjectMeta{
-					Name:      "tcp-service",
-					Namespace: "default",
-				},
-				Spec: v1.ServiceSpec{
-					Ports: []v1.ServicePort{{
-						Port:        80,
-						Protocol:    v1.ProtocolTCP,
-						TargetPort:  intstr.Parse("22"),
-						AppProtocol: &protocol,
-					}},
-					IPFamilies: []v1.IPFamily{v1.IPv4Protocol},
-					Type:       v1.ServiceTypeLoadBalancer,
-				},
-			}
+			oldSvc := newService("tcp-service", v1.ServiceTypeLoadBalancer, tweakAddPorts(v1.ProtocolTCP, 22), tweakAddAppProtocol("http"), tweakAddIPFamilies([]v1.IPFamily{v1.IPv4Protocol}))
```

Comment:
you said `:=` when you meant `=`.  I had to go look at this test - this is a very weird pattern.  If you wanbt to make this test really safer, change it so that the `updateFn` returns old and new values, rather than a function-scope variable that gets re-initialized in each case